### ISSUE TITLE
Some variable handlers misfire

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#5883]
As reported by @mojoTX on gitter (and bisected to 16fd7804846c0aeca926f207eb0e2ae86f819931), this way of setting key bindings leads to a non-functional shell:

```fish
function jarvis-vi-key-bindings
    fish_vi_key_bindings 
    for mode in default insert visual 
        bind -M $mode \cf forward-char 
    end 
end

set -g fish_key_bindings jarvis-vi-key-bindings
```

What happens: When fish_vi_key_bindings is called, it sets $fish_key_bindings to "fish_vi_key_bindings" to keep it in sync. Only that value seems to not stick for some reason, and since currently the vi-binding function erases the bindings before setting the variable it ends up not binding anything but those "forward-char"s, so you can't actually do anything.

Simplified:

```fish
function reload_var  --on-variable var
    if test "$var" = "$active_var" -a -n "$var"
        return
    end
    set -g active_var "$var"
    $var
end

function one
    if "$var" != one
        echo setting back to one
        set -g var one # <--- This does not work - the handler is triggered, but the value stays "two".
        return
    end
    echo one
end

function two
    one
    echo two
end

set -g var two
```

**Comment by faho** [comment]
Okay, it's some parser weirdness. `event_fire_internal` acquires the principal parser, which for some reason gets the old value?

**Comment by ridiculousfish** [comment]
Awesome investigation, this must have been painful to track down

**Comment by ridiculousfish** [comment]
The reduction has a line `if "$var" != one` which I think is missing a `test`. When I add that it seems to work like I expect. I'm still investigating.

**Comment by ridiculousfish** [comment]
The problem occurs when there's a global and universal variable with the same name and then you try to modify that variable without specifying a scope. It should modify the global, but in 16fd7804846c0aeca926f207eb0e2ae86f819931 it modifies the universal instead. Fixed in 292475148850512fd0649b7e83862348ff11b448
